src/components/chatbox/message.tsx

```diff
@@ -0,0 +1,317 @@
+import { IconWand24 } from '@hashicorp/flight-icons/svg-react/wand-24'
+
+import Button from 'components/button'
+import Text from 'components/text'
+import FeedbackForm from 'components/feedback-form'
+import IconTile from 'components/icon-tile'
+import useAuthentication from 'hooks/use-authentication'
+
+import s from './message.module.css'
+
+const UserMessage = ({ text, image }: { text: string; image?: string }) => {
+	return (
+		<div className={classNames(s.message, s.messageUser)}>
+			<div className={classNames(s.avatar)}>
+				{image ? <img src={image} alt="user avatar" /> : null}
+			</div>
+			<Text /* Body/200/Medium */
+				size={200}
+				weight="medium"
+				className={classNames(s.content, s.userInput)}
+			>
+				{text}
+			</Text>
+			<div className={classNames(s.gutter)}></div>
+		</div>
+	)
+}
+
+const AssistantMessage = ({
+	markdown,
+	showActions,
+	conversationId,
+	messageId,
+}: {
+	markdown: string
+	showActions: boolean
+	conversationId: string
+	messageId: string
+}) => {
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+	// Determines green/red button
+	const [rating, setRating] = useState<1 | -1 | 0>(0)
+
+	const handleFeedback = async ({
+		rating,
+		text,
+	}: {
+		rating?: -1 | 1
+		text?: string
+	}) => {
+		const response = await fetch('/api/chat/feedback', {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/json',
+				Authorization: `Bearer ${accessToken}`,
+			},
+			body: JSON.stringify({
+				messageId,
+				conversationId,
+				rating,
+				text,
+			}),
+		})
+		return response
+	}
+
+	const [isCopied, setIsCopied] = useState(false)
+	useEffect(() => {
+		let id: ReturnType<typeof setTimeout>
+		if (isCopied) {
+			id = setTimeout(() => {
+				setIsCopied(false)
+			}, 4000)
+		}
+		return () => {
+			clearTimeout(id)
+		}
+	}, [isCopied])
+
+	/** copy rich text; not raw markdown */
+	const handleCopy = async () => {
+		// CSS Module classname is a reasonable "hook" to grab the element
+		// rendered by ReactMarkdown.
+		const el = document.getElementsByClassName(s.markdown)?.[0]
+
+		if (el) {
+			// https://www.nikouusitalo.com/blog/why-isnt-clipboard-write-copying-my-richtext-html/
+			const clipboardItem = new ClipboardItem({
+				// @ts-expect-error - innerText is available
+				'text/plain': new Blob([el.innerText], { type: 'text/plain' }),
+				'text/html': new Blob([el.outerHTML], { type: 'text/html' }),
+			})
+
+			try {
+				await navigator.clipboard.write([clipboardItem])
+				setIsCopied(true)
+			} catch (err) {
+				// noop
+			}
+		}
+	}
+
+	return (
+		<div className={classNames(s.message, s.assistant)}>
+			<IconTile className={classNames(s.icon)}>
+				<IconWand24 style={{ width: 24, height: 24 }} />
+			</IconTile>
+			<div className={classNames(s.content)}>
+				<ReactMarkdown
+					className={s.markdown}
+					components={{
+						// @ts-expect-error - ignore this type incompatibility
```

Comment:
⛏️ Nitpick: this is really outside the scope of this work, so totally optional, but if you hop over to `mdx-code-blocks/index.tsx` and drop in a placeholder type for `MdxPre` I think you could drop this `@ts-expect-error`:

```tsx
export function MdxPre({
	className,
	...restProps
}: {
	/**
	 * @TODO(zchsh): `@hashicorp/react-code-block` needs some type work,
	 * this is a temporary solution for compatibility with `PreProps`
	 * in that package.
	 */
	children: $TSFixMe
	className?: string
	hasBarAbove?: boolean
	theme?: 'light' | 'dark'
}) {
```

Reply:
f52efdc8a - this is unfortunately more lines of code, but it's a large improvement with clear and immediately actionable steps.